### DISCUSSION TITLE
Suggestion - a retry hook utility for testing

### DISCUSSION THREAD

**Discussion by shahar4499** [celery/celery#7604] [Q&A]
If this seems to be useful, I will try to implement it, I already have a semi working util.

The idea is this:
Sometimes a task raises a retry, hoping that the next time it runs, it will succeed.
Example pseudo code scenario:

```
@app.task
def save_things_to_disk(disk, things):
    if not disk.enough_space(things):
        raise save_things_to_disk.retry()
    save(things)
```

This task hopes that on one of the retries, the disk will have enough space.
I would want to test the task flow like this (sort of):

```
def test_save_to_disk_retry_once_and_succeed():
    ...

    assert disk.is_full()
    save_things_to_disk.before_retry_hook(
        action=disk.make_space,
        kwargs={'MB': size},
        after_retries=1,
    )
    assert not disk.is_full()
    result = save_things_to_disk.delay().wait()
    assert success(result)

```

So what helps me here is the `before_retry_hook`, where I inject any functions and arguments to execute before the retry.
This can also come as a separate utility, not necessarily a method in Task.

**Comment by georgepsarakis** [comment]
@shahar4499 I think what you describe can be accomplished using [error callbacks](http://docs.celeryproject.org/en/latest/userguide/calling.html#linking-callbacks-errbacks).

**Comment by shahar4499** [comment]
@georgepsarakis Will the callbacks work when running the tasks in eager mode?

**Comment by georgepsarakis** [comment]
@shahar4499 I believe so, yes.
